File: src/CrimeWatch.Domain/AggregateModels/ModeratorAggregate/Moderator.cs
﻿using CrimeWatch.Domain.AggregateModels.AccountAggregate;
using CrimeWatch.Domain.AggregateModels.UserAggregate;

namespace CrimeWatch.Domain.AggregateModels.ModeratorAggregate;
public class Moderator : AggregateRoot<ModeratorId>
{
    public string PoliceId { get; set; } = string.Empty;
    public UserId UserId { get; set; } = default!;
    public AccountId AccountId { get; set; } = default!;
    public string Province { get; set; } = string.Empty;

    public User? User { get; set; }
    public Account? Account { get; set; }

    public static Moderator? Create(
            string firstName,
            string lastName,
            Gender gender,
            DateTime dateOfBirth,
            string phoneNumber,
            string policeId,
            string province,
            string email,
            string password
        )
    {
        var user = User.Create(firstName, lastName, gender, dateOfBirth, phoneNumber);
        var account = Account.Create(email, password, true);

        return new()
        {
            Id = new(Guid.NewGuid()),
            User = user,
            Account = account,
            PoliceId = policeId,
            Province = province
        };
    }

    public Moderator Update(
        string firstName,
        string lastName,
        Gender gender,
        DateTime dateOfBirth,
        string phoneNumber,
        string policeId,
        string province,
        string email,
        string password)
    {
        User!.FirstName = firstName;
        User!.LastName = lastName;
        User!.DateOfBirth = dateOfBirth;
        User!.PhoneNumber = phoneNumber;
        User!.Gender = gender;
        Account!.Email = email;
        Account!.Password = password;
        PoliceId = policeId;
        Province = province;

        return this;
    }
}

File: src/crimewatch.web.client/src/models/Moderator.ts
import { Gender } from "../enums/Gender";
import AccountId from "../valueObjects/AccountId";
import ModeratorId from "../valueObjects/ModeratorId";
import UserId from "../valueObjects/UserId";
import Account from "./Account";
import User from "./User";

export interface UpdateModeratorDto {
    id: ModeratorId;
    firstName: string;
    lastName: string;
    gender: Gender;
    dateOfBirth: Date;
    phoneNumber: string;
    policeId: string;
    province: string;
    email: string;
    password: string;
}


export interface CreateModeratorDto {
    firstName: string;
    lastName: string;
    gender: Gender;
    birthDate: Date;
    phoneNumber: string;
    email: string;
    password: string;
    policeId: string;
    province: string;
}

export default interface Moderator{
    moderatorId: ModeratorId;
    policeId: string;
    userId: UserId;
    accountId: AccountId;
    province: string;
    user: User | null;
    account: Account | null;
}